### ISSUE TITLE
[Feature]: Support sparse in-place weight updates in weight transfer API

### ISSUE THREAD

**Issue by qgallouedec** [vllm-project/vllm#39451]
### 🚀 The feature, motivation and pitch

In online RL, the trainer periodically syncs updated weights to a vLLM inference server. After a single optimizer step, typically >99% of bf16 elements are unchanged. We'd like to transfer and apply only the delta.

### Problem

`receive_weights` operates on full dense tensors. For each parameter, it allocates the full shape and broadcasts the entire tensor:

```python
# nccl_engine.py — receive_weights
for name, dtype_name, shape in zip(update_info.names, update_info.dtype_names, update_info.shapes):
    weight = torch.empty(shape, dtype=dtype, device="cuda")
    self.model_update_group.broadcast(weight, src=0)
    load_weights([(name, weight)])
```

There's no way to say "here are the 0.3% of elements that changed — apply them in-place." This forces us to keep a full CPU bf16 snapshot on the vLLM side to reconstruct dense tensors from sparse patches before calling `load_weights`.

### Possible API

A sparse variant that broadcasts only indices + values, then scatters directly into the existing GPU parameter:

```python
# Sparse path — broadcast only the delta
indices = torch.empty(nnz, dtype=torch.int32, device="cuda")
values = torch.empty(nnz, dtype=dtype, device="cuda")
self.model_update_group.broadcast(indices, src=0)
self.model_update_group.broadcast(values, src=0)
param = model.get_parameter(name)
param.flatten()[indices.long()] = values
```

This would eliminate the CPU snapshot and reduce the data transferred/copied from `O(numel)` to `O(nnz)` per parameter.

### Alternatives

Keep as is, it works, but it's suboptimal

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by mgoin** [comment]
What about using torch.sparse Tensors directly? https://docs.pytorch.org/docs/stable/sparse.html

I think this is a great idea to try!

**Comment by kylesayrs** [comment]
Great finding! We should consider first implementing this as a `sparse` flag which is non-composable with `packed` and `is_checkpoint_format`, then composing with `packed` later. This technique cannot compose with `is_checkpoint_format`, as it assumes that the kernel weight itself as sparse, not the checkpoint format weight.

**Comment by bedeks** [comment]
@qgallouedec @mgoin @kylesayrs I opened an MVP implementation for sparse NCCL weight transfer here: https://github.com/vllm-project/vllm/pull/40096

I also ran temporary repro/debug harnesses on a 2-GPU
pod to validate dense-vs-sparse equivalence for the same deterministic patch on
`Qwen/Qwen3-1.7B`.

Observed results:

- trainer patch digests matched
- full server-side parameter digest maps matched
- controlled `max_tokens=1`, greedy outputs matched between dense and sparse updates


Performance validation:

- for a patch affecting ~0.3% of model elements on Qwen/Qwen3-1.7B, the sparse payload was ~30.97 MB versus ~3.44 GB for the dense full-model resend path
- in one pod validation run, trainer-side send time decreased from ~175 ms for dense resend to ~4 ms for sparse patch transfer

Could you take a look and let me know if the approach/scope makes sense for the MVP? If it does, could you also add the `ready` label so CI can run?

**Comment by AmineDiro** [comment]
Hey, trl is adopting this new feature  in [AsyncGRPO](https://github.com/huggingface/trl/pull/5937) 🎉

We are now using sparse in-place weight updates in TRL's experimental AsyncGRPO:[ huggingface/trl#5937 ](https://github.com/huggingface/trl/pull/5937). After each optimizer step, we detect which bf16 elements actually changed, encode only those as a sparse patch, and apply them in place on the vLLM side using the sparse weight patch. 

Results (Qwen3-1.7B, GSM8K, single GPU):
- sparse patches apply end-to-end with 0 failures
- reward improves 0.10 → 0.50 through the delta path, so the model is learning
- payload and send time drop by orders of magnitude vs a full dense resend, matching the numbers reported

Thanks @bedeks and @hao-aaron, this is exactly the primitive we needed 👏🏼 

A few questions as we push it further, as we will probably work on having the sparse weight sync be the default path to do training in `trl`.
1. What is the concrete blocker for TP>1 / PP>1? 
2. If the feature compatible with expert parallelism and/or fused-MoE serving?
3. We are running this using V1 backen but will this feature work with  V2 runner?
